From 1973fa0e74900afc9384369730fa8e30815b8bd8 Mon Sep 17 00:00:00 2001
From: Anson Huang <Anson.Huang@nxp.com>
Date: Tue, 7 Aug 2018 15:43:34 +0800
Subject: [PATCH 4306/5242] MLK-19149 ARM: imx: fix low bus mode hang on
 i.MX7D

commit  c0bcdf73cec2a3d3966c63ce486a9f8fd323e0e9 from
https://source.codeaurora.org/external/imx/linux-imx.git

Per design requirement, AHB clock parent switch and divider
change needs to keep previous/current parent enabled but
when we switch the clock parent, previous AHB clock parent
may be disabled by common clock framework if the use count
is 0, so here we have to make sure AHB's previous parent
pfd2_270m is enabled during AHB set rate.

Signed-off-by: Anson Huang <Anson.Huang@nxp.com>
Reviewed-by: Robin Gong <yibin.gong@nxp.com>
(cherry picked from commit 41cb188e5f4732c7fdb83894399c4dc1303fd774)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/mach-imx/busfreq-imx.c |   11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/arch/arm/mach-imx/busfreq-imx.c b/arch/arm/mach-imx/busfreq-imx.c
index 2cbc033..b6aa072 100644
--- a/arch/arm/mach-imx/busfreq-imx.c
+++ b/arch/arm/mach-imx/busfreq-imx.c
@@ -618,6 +618,16 @@ static void exit_lpm_imx6sl(void)
 
 static void enter_lpm_imx7d(void)
 {
+	/*
+	 * The AHB clock parent switch and divider change
+	 * needs to keep previous/current parent enabled
+	 * per design requirement, but when we switch the
+	 * clock parent, previous AHB clock parent may be
+	 * disabled by common clock framework, so here we
+	 * have to make sure AHB's previous parent pfd2_270m
+	 * is enabled during AHB set rate.
+	 */
+	clk_prepare_enable(pfd2_270m);
 	if (audio_bus_count) {
 		clk_prepare_enable(pfd0_392m);
 		update_ddr_freq_imx_smp(HIGH_AUDIO_CLK);
@@ -647,6 +657,7 @@ static void enter_lpm_imx7d(void)
 		audio_bus_freq_mode = 0;
 		cur_bus_freq_mode = BUS_FREQ_LOW;
 	}
+	clk_disable_unprepare(pfd2_270m);
 }
 
 static void exit_lpm_imx7d(void)
-- 
1.7.9.5

